### PR TITLE
stargate: split rpc/

### PR DESCRIPTION
**Pull request by fedekunze**
- stargate: split proto and base files
- stargate: split app/ + codec/
- stargate: split types/
- stargate: split crypto/
- stargate: split x/evm/types/
- stargate: split x/evm/ (minus types/)
- stargate: split rpc/

### CODE REVIEW

File: rpc/backend/backend.go [hunk 1]
```diff
@@ -43,155 +42,114 @@ var _ Backend = (*EthermintBackend)(nil)
...
 	}
 
-	return hexutil.Uint64(blockNumber), nil
+	return hexutil.Uint64(info.LastHeight), nil
 }
 
 // GetBlockByNumber returns the block identified by number.
 func (b *EthermintBackend) GetBlockByNumber(blockNum rpctypes.BlockNumber, fullTx bool) (map[string]interface{}, error) {
-	height := blockNum.Int64()
-	if height <= 0 {
```
Comment (noot) [inline rpc/backend/backend.go:75]: the case where `blockNum == PendingBlockNumber` isn't handled anymore

File: rpc/config.go [hunk 2]
```diff
@@ -1,6 +1,14 @@
...
 
 import (
+<<<<<<< HEAD
+	"github.com/gorilla/mux"
+
+	"github.com/cosmos/cosmos-sdk/client"
+
+	"github.com/cosmos/ethermint/rpc/websockets"
+	"github.com/cosmos/ethermint/server/config"
+=======
```
Comment (noot) [inline rpc/config.go]: fix merge conflicts

File: rpc/config.go [hunk 3]
```diff
@@ -67,6 +48,8 @@ func RegisterRoutes(rs *lcd.RestServer) {
 			panic(err)
 		}
 	}
+<<<<<<< HEAD
+=======
```
Comment (noot) [inline rpc/config.go]: same

File: rpc/namespaces/eth/api.go [hunk 4]
```diff
@@ -575,15 +571,22 @@ func (api *PublicEthereumAPI) doCall(
...
-	if args.To != nil {
-		toAddr = sdk.AccAddress(args.To.Bytes())
+	var fromAddr sdk.AccAddress
+	if args.From != nil {
+		fromAddr = sdk.AccAddress(args.From.Bytes())
+	}
+
+	accNum, seq, err := api.clientCtx.AccountRetriever.GetAccountNumberSequence(api.clientCtx, fromAddr)
+	if err != nil {
+		return nil, err
```
Comment (noot) [inline rpc/namespaces/eth/api.go:581]: if `args.From == nil` then an empty address gets passed here

File: rpc/namespaces/eth/api.go [hunk 5]
```diff
@@ -575,15 +571,22 @@ func (api *PublicEthereumAPI) doCall(
...
 	}
 
 	var msgs []sdk.Msg
 	// Create new call message
-	msg := evmtypes.NewMsgEthermint(0, &toAddr, sdk.NewIntFromBigInt(value), gas,
-		sdk.NewIntFromBigInt(gasPrice), data, sdk.AccAddress(addr.Bytes()))
+	msg := evmtypes.NewMsgEthereumTx(seq, args.To, value, gas, gasPrice, data)
```
Comment (noot) [inline rpc/namespaces/eth/api.go:586]: this has no `from` addr, does that mean it's a `call`? how does the signature check pass?

File: rpc/types/utils.go [hunk 6]
```diff
@@ -208,10 +252,11 @@ func GetBlockCumulativeGas(cdc *codec.Codec, block *tmtypes.Block, idx int) uint
 		}
 
 		switch tx := txi.(type) {
-		case authtypes.StdTx:
-			gasUsed += tx.GetGas()
-		case evmtypes.MsgEthereumTx:
+		case *evmtypes.MsgEthereumTx:
 			gasUsed += tx.GetGas()
+			// case authtypes.:
+			// 	gasUsed += tx.GetGas()
```
Comment (noot) [inline rpc/types/utils.go:258]: remove?

**Review by noot — COMMENTED** [review]
overall looks good! just a question about handling pending height

File: rpc/config.go [hunk 7]
```diff
@@ -20,45 +28,18 @@ import (
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/spf13/viper"
 )
+>>>>>>> 9ecd264ae0e8c6ace10abe3bc45ab8d3c29dc10f
```
Comment (freddyli7) [inline rpc/config.go:31]: same here lol

File: rpc/backend/backend.go [hunk 8]
```diff
@@ -43,155 +42,126 @@ var _ Backend = (*EthermintBackend)(nil)
...
 type EthermintBackend struct {
-	ctx       context.Context
-	clientCtx clientcontext.CLIContext
-	logger    log.Logger
-	gasLimit  int64
+	ctx         context.Context
+	clientCtx   client.Context
+	queryClient *rpctypes.QueryClient // gRPC query client
+	logger      log.Logger
+	gasLimit    int64
```
Comment (araskachoi) [inline rpc/backend/backend.go:49]: does this describe the block gas limit?

Comment (fedekunze) [reply rpc/backend/backend.go:49]: idk why is the gas limit a field here. Will update

**Pull request closed without merging**